packages/blocks/src/api/raw-handling/special-comment-converter.js

```diff
@@ -30,28 +29,79 @@ export default function specialCommentConverter( node, doc ) {
 	}
 
 	if ( node.nodeValue.indexOf( 'more' ) === 0 ) {
-		// Grab any custom text in the comment.
-		const customText = node.nodeValue.slice( 4 ).trim();
+		moreCommentConverter( node, doc );
+	}
+}
 
-		/*
-		 * When a `<!--more-->` comment is found, we need to look for any
-		 * `<!--noteaser-->` sibling, but it may not be a direct sibling
-		 * (whitespace typically lies in between)
-		 */
-		let sibling = node;
-		let noTeaser = false;
-		while ( ( sibling = sibling.nextSibling ) ) {
-			if (
-				sibling.nodeType === sibling.COMMENT_NODE &&
-				sibling.nodeValue === 'noteaser'
-			) {
-				noTeaser = true;
-				remove( sibling );
-				break;
-			}
+/**
+ * Convert `<!--more-->` as well as the `<!--more Some text-->` variant
+ * and its `<!--noteaser-->` companion into the custom element
+ * described in `specialCommentConverter()`.
+ *
+ * @param {Node}     node The node to be processed.
+ * @param {Document} doc  The document of the node.
+ * @return {void}
+ */
+function moreCommentConverter( node, doc ) {
+	// Grab any custom text in the comment.
+	const customText = node.nodeValue.slice( 4 ).trim();
+
+	/*
+	 * When a `<!--more-->` comment is found, we need to look for any
+	 * `<!--noteaser-->` sibling, but it may not be a direct sibling
+	 * (whitespace typically lies in between)
+	 */
+	let sibling = node;
+	let noTeaser = false;
+	while ( ( sibling = sibling.nextSibling ) ) {
+		if (
+			sibling.nodeType === sibling.COMMENT_NODE &&
+			sibling.nodeValue === 'noteaser'
+		) {
+			noTeaser = true;
+			remove( sibling );
+			break;
 		}
+	}
+
+	const moreBlock = createMore( customText, noTeaser, doc );
+
+	// If our `<!--more-->` comment is in the middle of a paragraph, we should
+	// split the paragraph in two and insert the more block in between. If not,
+	// the more block will eventually end up being inserted after the paragraph.
+	if (
+		! node.parentNode ||
+		node.parentNode.nodeName !== 'P' ||
+		node.parentNode.childNodes.length === 1
+	) {
+		replace( node, moreBlock );
+	} else {
+		const childNodes = Array.from( node.parentNode.childNodes );
+		const nodeIndex = childNodes.indexOf( node );
+		const wrapperNode = node.parentNode.parentNode || doc.body;
+
+		const paragraphBuilder = ( acc, child ) => {
+			if ( ! acc ) {
+				acc = doc.createElement( 'p' );
+			}
+
+			acc.appendChild( child );
+
+			return acc;
+		};
+
+		// Split the original parent node and insert our more block
+		[
+			childNodes.slice( 0, nodeIndex ).reduce( paragraphBuilder, null ),
+			moreBlock,
+			childNodes.slice( nodeIndex + 1 ).reduce( paragraphBuilder, null ),
+		].forEach(
```

Comment:
Do folks know if we've polyfilled https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith ? It's 💯 optional, but might be nice here.

eg
```js
paragraphWithMore.replaceWith( firstParagraph, moreBlock, secondParagraph );
```